backend/provisioner/dev/dev_provisioner.go

```diff
@@ -0,0 +1,187 @@
+	panic("unimplemented")
+}
+
+func (d *Provisioner) Provision(ctx context.Context, req *connect.Request[provisioner.ProvisionRequest]) (*connect.Response[provisioner.ProvisionResponse], error) {
+	if d.running == nil {
+		d.running = map[string]*task{}
+	}
+
+	previous := map[string]*provisioner.Resource{}
+	for _, r := range req.Msg.ExistingResources {
+		previous[r.ResourceId] = r
+	}
+
+	task := &task{}
+	for _, r := range req.Msg.DesiredResources {
+		if _, ok := previous[r.Resource.ResourceId]; !ok {
+			switch tr := r.Resource.Resource.(type) {
+			case *provisioner.Resource_Postgres:
+				step := &step{resource: r.Resource}
+				task.steps = append(task.steps, step)
+
+				d.provisionPostgres(ctx, tr, req.Msg.Module, r.Resource.ResourceId, step)
+			default:
+			}
+		}
+	}
+
+	if len(task.steps) == 0 {
+		return connect.NewResponse(&provisioner.ProvisionResponse{
+			Status: provisioner.ProvisionResponse_NO_CHANGES,
+		}), nil
+	}
+
+	token := strconv.Itoa(rand.Int()) //nolint:gosec
+	d.running[token] = task
+
+	return connect.NewResponse(&provisioner.ProvisionResponse{
+		ProvisioningToken: token,
+		Status:            provisioner.ProvisionResponse_SUBMITTED,
+	}), nil
+}
+
+func (d *Provisioner) Status(ctx context.Context, req *connect.Request[provisioner.StatusRequest]) (*connect.Response[provisioner.StatusResponse], error) {
+	token := req.Msg.ProvisioningToken
+	task, ok := d.running[token]
+	if !ok {
+		return statusFailure("unknown token")
+	}
+	if !task.Done() {
+		return connect.NewResponse(&provisioner.StatusResponse{
+			Status: &provisioner.StatusResponse_Running{},
+		}), nil
+	}
+
+	var resources []*provisioner.Resource
+	for _, step := range task.steps {
+		if step.err != nil {
+			return statusFailure(step.err.Error())
+		}
+		resources = append(resources, step.resource)
+	}
+	delete(d.running, token)
+
+	return connect.NewResponse(&provisioner.StatusResponse{
+		Status: &provisioner.StatusResponse_Success{
+			Success: &provisioner.StatusResponse_ProvisioningSuccess{
+				UpdatedResources: resources,
+			},
+		},
+	}), nil
+}
+
+func statusFailure(message string) (*connect.Response[provisioner.StatusResponse], error) {
+	return connect.NewResponse(&provisioner.StatusResponse{
+		Status: &provisioner.StatusResponse_Failed{
+			Failed: &provisioner.StatusResponse_ProvisioningFailed{
+				ErrorMessage: message,
+			},
+		},
+	}), nil
+}
+
+func (d *Provisioner) provisionPostgres(ctx context.Context, tr *provisioner.Resource_Postgres, module, id string, step *step) {
+	logger := log.FromContext(ctx)
+	logger.Infof("provisioning postgres database: %s_%s", module, id)
+
+	go func() {
+		defer step.done.Store(true)
+		if d.postgresDSN == "" {
+			dsn, err := dev.SetupDB(ctx, d.postgresImage, d.postgresPort, false)
+			if err != nil {
+				step.err = err
+				return
+			}
+			d.postgresDSN = dsn
+		}
+		dbName := module + "_" + id
```

Comment:
The ID will likely be camelCase, so you might want to use `strcase.LowerSnakeCase()` to canonicalise it.